src/functional/iterator/embedded.py

```diff
@@ -1103,16 +1111,26 @@ def closure(
         out = as_tuple_field(out) if can_be_tuple_field(out) else out
 
         for pos in _domain_iterator(domain):
+            promoted_ins = []
+            for inp in ins:
+                if isinstance(inp, LocatedField) or isinstance(inp, tuple):
+                    promoted_ins.append(inp)
+                else:
+                    inp_type = xtyping.infer_type(inp)
+                    if np.issubdtype(inp_type, np.number):
+                        promoted_ins.append(constant_field(inp))
+                    else:
+                        raise ValueError(
+                            "Expected a `Field` or a number (`float`, `np.int64`, ...)."
+                        )
```

Comment:
```python
Scalar: TypeAlias = SupportsInt | SupportsFloat | np.int32 | np.int64 | np.float32 | np.float64
CompositeOfScalarOrField = Scalar | LocatedField | tuple["CompositeOfScalarOrField", ...]
def promote_scalars(val: CompositeOfScalarOrField):
    """Given a scalar, field or composite thereof promote all (contained) scalars to fields."""
    if isinstance(val, tuple):
        return tuple(promote_scalars(el) for el in val)
    val_type = xtyping.infer_type(val)
    if np.issubdtype(val_type, np.number):
        promoted_ins.append(constant_field(val))
    else:
        raise ValueError(
            f"Expected a `Field` or a number (`float`, `np.int64`, ...), but got `val_type`."
        )
```
I just recognized this fails for nested tuples. Please add this (recursive) function at the top of the file (type aliases to the other ones) then you can simply do this:
```suggestion
            promoted_ins = [promote_scalars(inp) for inp in ins]
```
Salami slicing my own proposals now...